lib/Basic/Unix/TaskQueue.inc

```diff
@@ -320,164 +321,320 @@ void TaskQueue::addTask(const char *ExecPath, ArrayRef<const char *> Args,
   QueuedTasks.push(std::move(T));
 }
 
-bool TaskQueue::execute(TaskBeganCallback Began, TaskFinishedCallback Finished,
-                        TaskSignalledCallback Signalled) {
+// Owns Tasks, handles correspondence with FDs and pids.
+// FIXME: only handles stdout fds.
+class TaskMap {
+  using PidToTaskMap = llvm::DenseMap<pid_t, std::unique_ptr<Task>>;
+  PidToTaskMap TasksByPid;
+
+public:
+  TaskMap() = default;
+
+  bool empty() const { return TasksByPid.empty(); }
+  unsigned size() const { return TasksByPid.size(); }
+
+  void add(std::unique_ptr<Task> T) { TasksByPid[T->getPid()] = std::move(T); }
+
+  Task &findTaskForFd(const int fd) {
+    auto predicate = [&fd](PidToTaskMap::value_type &value) -> bool {
+      return value.second->getPipe() == fd;
+    };
+    auto iter = std::find_if(TasksByPid.begin(), TasksByPid.end(), predicate);
+    assert(iter != TasksByPid.end() &&
+           "All outstanding fds must be associated with a  Task");
+    return *iter->second;
+  }
+
+  void eraseTask(Task &T) { TasksByPid.erase(T.getPid()); }
+};
+
+/// Concurrently execute the tasks in the TaskQueue, collecting the outputs from
+/// each task. The typical task is a Swift frontend job.
+/// Maintain invarients connecting tasks to execute, tasks currently executing,
+/// and fds being polled. Handle output from and death of tasks.
+class TaskShepherd {
```

Comment:
Can you make it clearer why the TaskShepherd is separate from the TaskQueue at all?

Reply:
I didn't want to mess with the basic version of the TaskQueue, but I suppose I could have. There could be an abstract superclass, and two concrete subclasses, one for Unix and the other for the other cases. Instead, I took the function that was there, and reified it as a class.

How could the separation be made clearer? I could spell out the invarients maintained by this class. Would that help?